In [1]:
import pandas as pd
import numpy as np
import warnings
warnings.filterwarnings('ignore')

## **Q1**

In [2]:
df_ipl = pd.read_csv('deliveries.csv')
df_ipl.info() # checking purpose

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 179078 entries, 0 to 179077
Data columns (total 21 columns):
 #   Column            Non-Null Count   Dtype 
---  ------            --------------   ----- 
 0   match_id          179078 non-null  int64 
 1   inning            179078 non-null  int64 
 2   batting_team      179078 non-null  object
 3   bowling_team      179078 non-null  object
 4   over              179078 non-null  int64 
 5   ball              179078 non-null  int64 
 6   batsman           179078 non-null  object
 7   non_striker       179078 non-null  object
 8   bowler            179078 non-null  object
 9   is_super_over     179078 non-null  int64 
 10  wide_runs         179078 non-null  int64 
 11  bye_runs          179078 non-null  int64 
 12  legbye_runs       179078 non-null  int64 
 13  noball_runs       179078 non-null  int64 
 14  penalty_runs      179078 non-null  int64 
 15  batsman_runs      179078 non-null  int64 
 16  extra_runs        179078 non-null  int

In [ ]:
# checking purpose
df_ipl.head(2)

match_id  inning         batting_team                 bowling_team  over  \
0         1       1  Sunrisers Hyderabad  Royal Challengers Bangalore     1   
1         1       1  Sunrisers Hyderabad  Royal Challengers Bangalore     1   

   ball    batsman non_striker    bowler  is_super_over  ...  bye_runs  \
0     1  DA Warner    S Dhawan  TS Mills              0  ...         0   
1     2  DA Warner    S Dhawan  TS Mills              0  ...         0   

   legbye_runs  noball_runs  penalty_runs  batsman_runs  extra_runs  \
0            0            0             0             0           0   
1            0            0             0             0           0   

   total_runs  player_dismissed dismissal_kind fielder  
0           0               NaN            NaN     NaN  
1           0               NaN            NaN     NaN  

[2 rows x 21 columns]

In [ ]:
# checking dismissal kind categories along with number of dismissals in each
df_ipl[['dismissal_kind']].value_counts('dismissal_kind')

dismissal_kind
caught                   5348
bowled                   1581
run out                   852
lbw                       540
stumped                   278
caught and bowled         211
retired hurt               12
hit wicket                 10
obstructing the field       2
Name: count, dtype: int64

> we must exclude the categories obstructing the field, retired hurt and runout. because they are not counted into bowlers account.

In [ ]:
df_bowlers = df_ipl[(df_ipl['dismissal_kind'] != 'obstructing the field') & (df_ipl['dismissal_kind'] != 'retired hurt') & (df_ipl['dismissal_kind'] != 'run out')]

In [ ]:
df_bowlers = df_bowlers[df_bowlers['dismissal_kind'].notna()]
df_bowlers.head()

match_id  inning                 batting_team  \
11          1       1          Sunrisers Hyderabad   
64          1       1          Sunrisers Hyderabad   
94          1       1          Sunrisers Hyderabad   
116         1       1          Sunrisers Hyderabad   
160         1       2  Royal Challengers Bangalore   

                    bowling_team  over  ball        batsman   non_striker  \
11   Royal Challengers Bangalore     2     5      DA Warner      S Dhawan   
64   Royal Challengers Bangalore    11     3       S Dhawan  MC Henriques   
94   Royal Challengers Bangalore    16     2   MC Henriques  Yuvraj Singh   
116  Royal Challengers Bangalore    19     4   Yuvraj Singh      DJ Hooda   
160          Sunrisers Hyderabad     6     4  Mandeep Singh      CH Gayle   

          bowler  is_super_over  ...  bye_runs  legbye_runs  noball_runs  \
11   A Choudhary              0  ...         0            0            0   
64     STR Binny              0  ...         0            0            0   
94     YS Chahal              0  ...         0            0            0   
116     TS Mills              0  ...         0            0            0   
160  Rashid Khan              0  ...         0            0            0   

     penalty_runs  batsman_runs  extra_runs  total_runs  player_dismissed  \
11              0             0           0           0         DA Warner   
64              0             0           0           0          S Dhawan   
94              0             0           0           0      MC Henriques   
116             0             0           0           0      Yuvraj Singh   
160             0             0           0           0     Mandeep Singh   

    dismissal_kind        fielder  
11          caught  Mandeep Singh  
64          caught    Sachin Baby  
94          caught    Sachin Baby  
116         bowled            NaN  
160         bowled            NaN  

[5 rows x 21 columns]

In [ ]:
print('Below are top 5 bowlers who taken the most wickets across all matches:')
df_top_bowlers = df_bowlers.groupby('bowler')[['dismissal_kind']].count().sort_values(by='dismissal_kind',ascending=False).head(5)
df_top_bowlers.columns = ['Total Wickets']
df_top_bowlers

Below are top 5 bowlers who taken the most wickets across all matches:


Total Wickets
bowler                        
SL Malinga                 170
A Mishra                   156
Harbhajan Singh            150
PP Chawla                  149
DJ Bravo                   147

## **Q2**

In [ ]:
df_bowler_economy = df_ipl.groupby('bowler')[['total_runs', 'penalty_runs']].sum()
df_bowler_economy['runs_by_bowler'] = df_bowler_economy['total_runs'] - df_bowler_economy['penalty_runs']
df_bowler_economy['total_overs'] = df_ipl.groupby('bowler')[['ball']].count() / 6
df_bowler_economy['economy_rate'] = df_bowler_economy['runs_by_bowler'] / df_bowler_economy['total_overs']


In [ ]:
print('Bowlers with lowest economy rate:')
df_bowler_economy.sort_values(by='economy_rate').head(10)

Bowlers with lowest economy rate:


total_runs  penalty_runs  runs_by_bowler  total_overs  \
bowler                                                                 
AC Gilchrist            0             0               0     0.166667   
NB Singh               18             0              18     4.166667   
Sachin Baby             8             0               8     1.666667   
S Dube                  9             0               9     1.833333   
AM Rahane               5             0               5     1.000000   
LA Carseldine           6             0               6     1.166667   
SS Mundhe               6             0               6     1.166667   
DJ Thornely            40             0              40     7.333333   
A Roy                  14             0              14     2.500000   
M Manhas               42             0              42     7.000000   

               economy_rate  
bowler                       
AC Gilchrist       0.000000  
NB Singh           4.320000  
Sachin Baby        4.800000  
S Dube             4.909091  
AM Rahane          5.000000  
LA Carseldine      5.142857  
SS Mundhe          5.142857  
DJ Thornely        5.454545  
A Roy              5.600000  
M Manhas           6.000000

> But this will not give proper insight as we are not considered minimum number of matches. Because economy depends on number of matches also. Also there are situations like a batsman may bowl in a match then he never bowls in his career after that. Then that person actually doesn't comes under this category. <br>
> So here we should take number of matches category or minimum number of overs.

## **Q3**

In [ ]:
# method 1
df_dismissals_count = pd.DataFrame(df_ipl[df_ipl['dismissal_kind'].notna()].value_counts('dismissal_kind'))
df_dismissals_count.columns = ['total_Dismissals']
df_dismissals_count['perc_chance_of_dismissal'] = (df_dismissals_count['total_Dismissals'] / df_dismissals_count['total_Dismissals'].sum()) * 100
df_dismissals_count.head(5)

total_Dismissals  perc_chance_of_dismissal
dismissal_kind                                            
caught                      5348                 60.538827
bowled                      1581                 17.896763
run out                      852                  9.644555
lbw                          540                  6.112746
stumped                      278                  3.146932

In [ ]:
# method 2
df_dismissals_count = df_ipl.groupby('dismissal_kind')[['batsman']].count().sort_values(by='batsman',ascending=False)
df_dismissals_count.columns = ['total_Dismissals']
df_dismissals_count['perc_chance_of_dismissal'] = (df_dismissals_count['total_Dismissals'] / df_dismissals_count['total_Dismissals'].sum()) * 100
df_dismissals_count.head(5)

total_Dismissals  perc_chance_of_dismissal
dismissal_kind                                            
caught                      5348                 60.538827
bowled                      1581                 17.896763
run out                      852                  9.644555
lbw                          540                  6.112746
stumped                      278                  3.146932

> **Insight:** The most `common dismissal kind` for a batsman is `by caught`. Then bowled is next common dismissal.<br>
> - i.e. mostly a batsman may dismissed by catch. Next we can think of is by bowled.<br>
> - In an accurate there will be 60.5% chance a batsman may dismiss by catch. Then approx 18% chance by bowled.

## **Q4**

In [ ]:
df_batsman_runs = df_ipl.groupby('batsman').agg({'batsman_runs':['sum'], 'ball':['count']}).sort_values(by=('batsman_runs','sum'),ascending=False)
df_batsman_runs.columns = ['total_runs', 'balls_faced']

In [ ]:
df_batsman_runs = df_batsman_runs[df_batsman_runs['balls_faced'] >= 100]
df_batsman_runs['strike_rate'] = (df_batsman_runs['total_runs'] / df_batsman_runs['balls_faced']) * 100

In [ ]:
print('Top 5 batsmen with highest strike rate who faced minimum 100 balls:')
df_batsman_runs.sort_values(by='strike_rate',ascending=False).head(5)

Top 5 batsmen with highest strike rate who faced minimum 100 balls:


total_runs  balls_faced  strike_rate
batsman                                          
AD Russell         1445          803   179.950187
M Ali               311          183   169.945355
SP Narine           803          481   166.943867
BCJ Cutting         240          146   164.383562
RR Pant            1792         1104   162.318841

## **Q5**

In [ ]:
df_partnership = df_ipl.groupby(['match_id','batsman', 'non_striker'])[['total_runs']].sum().sort_values(by='total_runs',ascending=False)
df_partnership_rev = df_ipl.groupby(['match_id', 'non_striker', 'batsman'])[['total_runs']].sum().sort_values(by='total_runs',ascending=False)
df_partnership = pd.concat((df_partnership, df_partnership_rev))
df_partnership = df_partnership.groupby(df_partnership.index)[['total_runs']].sum().sort_values(by='total_runs',ascending=False)
df_partnership.drop_duplicates(inplace=True)
df_partnership[['match_id', 'batsman_1', 'batsman_2']] = pd.DataFrame(df_partnership.index.values.tolist(), index= df_partnership.index)
df_partnership.reset_index(drop=True, inplace=True)
df_partnership = df_partnership.reindex(columns=['match_id', 'batsman_1', 'batsman_2', 'total_runs'])

In [ ]:
print('Below are the details of top 5 highest partnership between 2 batsmen:')
df_partnership.head(5)

Below are the details of top 5 partnership between 2 batsmen:


match_id       batsman_1 batsman_2  total_runs
0       620  AB de Villiers   V Kohli         229
1       562  AB de Villiers   V Kohli         215
2       296    AC Gilchrist  SE Marsh         206
3       372         V Kohli  CH Gayle         204
4       331       DA Warner   NV Ojha         189

> **Insight:** Highest partnership  in IPL is 229 by V Kohli and AB de Villers. Even second highest also made by them with partnership runs of 215.

## **Q6**

In [3]:
df_powerplay = df_ipl[(df_ipl['over'] <= 6)]
df_powerplay_runs = df_powerplay.groupby(['match_id', 'batsman'])[['batsman_runs']].sum().sort_values(by='batsman_runs', ascending=False)
df_powerplay_runs.columns = ['runs_in_powerplay']

In [4]:
print('Below are the details of top 5 batsmen who scored highest runs in powerplay overs:')
df_powerplay_runs.head(5)

Below are the details of top 5 batsmen who scored highest runs in powerplay overs:


runs_in_powerplay
match_id batsman                        
516      SK Raina                     87
172      AC Gilchrist                 74
7925     JC Buttler                   68
36       DA Warner                    62
11346    KL Rahul                     59

## **Q7**

In [ ]:
df_extra_runs = df_ipl.groupby(['match_id', 'bowling_team'])[['extra_runs']].sum().sort_values(by='extra_runs', ascending=False)
print('Top 5 teams which conceded most extra runs in IPL:')
df_extra_runs.head(5)

Top 5 teams which conceded most extra runs in IPL:


extra_runs
match_id bowling_team                           
64       Deccan Chargers                      28
296      Royal Challengers Bangalore          27
570      Rajasthan Royals                     26
101      Royal Challengers Bangalore          26
162      Mumbai Indians                       26

## **Q8**

In [ ]:
df_last_5_overs = df_ipl[df_ipl['over'] >= 16]
df_last_5_overs_economy = df_last_5_overs.groupby(['match_id', 'bowler'])[['total_runs', 'penalty_runs']].sum()
df_last_5_overs_economy['runs_given'] = df_last_5_overs_economy['total_runs'] - df_last_5_overs_economy['penalty_runs']
df_last_5_overs_economy['total_overs'] = df_last_5_overs.groupby(['match_id', 'bowler'])[['ball']].count() / 6
df_last_5_overs_economy['economy_rate'] = df_last_5_overs_economy['runs_given'] / df_last_5_overs_economy['total_overs']
df_last_5_overs_economy['wickets'] = df_last_5_overs.groupby(['match_id','bowler'])[['dismissal_kind']].count()
df_last_5_overs_economy.drop(['total_runs', 'penalty_runs'], axis=1, inplace=True)

In [ ]:
print('Below are the top 5 bowlers with highest wickets along with best economy rate in last 5 overs:')
df_last_5_overs_economy.sort_values(by=['wickets', 'economy_rate'], ascending=[False, True]).head(5)

Below are the top 5 bowlers with highest wickets along with best economy rate in last 5 overs:


runs_given  total_overs  economy_rate  wickets
match_id bowler                                                    
616      A Zampa             14     3.000000      4.666667        6
43       JD Unadkat          10     2.000000      5.000000        5
7947     P Krishna           20     3.166667      6.315789        5
11322    SL Malinga          19     2.000000      9.500000        5
300      A Mishra             3     2.000000      1.500000        4

In [ ]:
print('Below are the top 5 bowlers with lowest economy rate with most wickets in last 5 overs:')
df_last_5_overs_economy.sort_values(by=['economy_rate', 'wickets'], ascending=[True, False]).head(5)

Below are the top 5 bowlers with lowest economy rate with most wickets in last 5 overs:


runs_given  total_overs  economy_rate  wickets
match_id bowler                                                    
132      SL Malinga           0     0.333333           0.0        2
202      SB Jakati            0     1.000000           0.0        2
233      A Kumble             0     0.500000           0.0        2
265      DW Steyn             0     0.500000           0.0        2
285      SK Raina             0     0.500000           0.0        2

## **Q9**

In [ ]:
print('Below are the details of top 5 fielders who caught most catches:')
df_ipl[df_ipl['dismissal_kind'] == 'caught'].value_counts('fielder').head(5)

Below are the details of top 5 fielders who caught most catches:


fielder
KD Karthik        109
SK Raina           99
MS Dhoni           98
AB de Villiers     93
RV Uthappa         84
Name: count, dtype: int64

## **Q10**

In [ ]:
df_innings = df_ipl.groupby(['match_id', 'inning', 'batting_team']).agg({'total_runs': 'sum', 'player_dismissed': 'count'}).reset_index()
df_innings.columns = ['match_id', 'inning', 'batting_team', 'total_runs', 'wickets']
df_innings['Score'] = df_innings['total_runs'].astype(str) + ' - ' + df_innings['wickets'].astype(str)
df_innings.head()

match_id  inning                 batting_team  total_runs  wickets  \
0         1       1          Sunrisers Hyderabad         207        4   
1         1       2  Royal Challengers Bangalore         172       10   
2         2       1               Mumbai Indians         184        8   
3         2       2       Rising Pune Supergiant         187        3   
4         3       1                Gujarat Lions         183        4   

      Score  
0   207 - 4  
1  172 - 10  
2   184 - 8  
3   187 - 3  
4   183 - 4

In [ ]:
df_innings_1 = df_innings[df_innings['inning'] == 1]
df_innings_2 = df_innings[df_innings['inning'] == 2]
df_innings = pd.merge(df_innings_1, df_innings_2, on=['match_id'], how='inner')
df_innings['win'] = [f"{i[2]}" if i[3] > i[8] else f'{i[7]}' for i in df_innings.values]
df_innings['win by runs'] = df_innings['total_runs_x'] - df_innings['total_runs_y']
df_innings.drop(['total_runs_x', 'wickets_x', 'total_runs_y', 'wickets_y'], axis=1, inplace=True)
df_innings.head()

match_id  inning_x               batting_team_x  Score_x  inning_y  \
0         1         1          Sunrisers Hyderabad  207 - 4         2   
1         2         1               Mumbai Indians  184 - 8         2   
2         3         1                Gujarat Lions  183 - 4         2   
3         4         1       Rising Pune Supergiant  163 - 6         2   
4         5         1  Royal Challengers Bangalore  157 - 8         2   

                batting_team_y   Score_y                          win  \
0  Royal Challengers Bangalore  172 - 10          Sunrisers Hyderabad   
1       Rising Pune Supergiant   187 - 3       Rising Pune Supergiant   
2        Kolkata Knight Riders   184 - 0        Kolkata Knight Riders   
3              Kings XI Punjab   164 - 4              Kings XI Punjab   
4             Delhi Daredevils   142 - 9  Royal Challengers Bangalore   

   win by runs  
0           35  
1           -3  
2           -1  
3           -1  
4           15

In [ ]:
df_chases = df_innings[df_innings['win by runs'] < 0]
df_successfull_chases = df_chases['batting_team_y'].value_counts().reset_index()
df_successfull_chases.columns = ['team', 'successfull_chases']
df_successfull_chases

team  successfull_chases
0         Kolkata Knight Riders                  51
1                Mumbai Indians                  48
2           Chennai Super Kings                  48
3   Royal Challengers Bangalore                  45
4              Rajasthan Royals                  42
5              Delhi Daredevils                  40
6               Kings XI Punjab                  37
7           Sunrisers Hyderabad                  24
8                 Gujarat Lions                  12
9               Deccan Chargers                  11
10                Pune Warriors                   6
11       Rising Pune Supergiant                   5
12         Kochi Tuskers Kerala                   4
13      Rising Pune Supergiants                   3
14               Delhi Capitals                   3

In [ ]:
df_total_chases = df_innings['batting_team_y'].value_counts().reset_index()
df_total_chases.columns = ['team', 'total_chases']
df_chases = pd.merge(df_total_chases, df_successfull_chases, on='team', how='inner')
df_chases['win_percentage'] = (df_chases['successfull_chases'] / df_chases['total_chases']) * 100
df_chases.sort_values(by='win_percentage', ascending=False).reset_index(drop=True)

team  total_chases  successfull_chases  \
0                 Gujarat Lions            16                  12   
1           Chennai Super Kings            75                  48   
2        Rising Pune Supergiant             8                   5   
3          Kochi Tuskers Kerala             7                   4   
4                Mumbai Indians            86                  48   
5         Kolkata Knight Riders            95                  51   
6           Sunrisers Hyderabad            45                  24   
7              Rajasthan Royals            79                  42   
8   Royal Challengers Bangalore            95                  45   
9              Delhi Daredevils            89                  40   
10              Kings XI Punjab            85                  37   
11      Rising Pune Supergiants             7                   3   
12              Deccan Chargers            32                  11   
13               Delhi Capitals            10                   3   
14                Pune Warriors            25                   6   

    win_percentage  
0        75.000000  
1        64.000000  
2        62.500000  
3        57.142857  
4        55.813953  
5        53.684211  
6        53.333333  
7        53.164557  
8        47.368421  
9        44.943820  
10       43.529412  
11       42.857143  
12       34.375000  
13       30.000000  
14       24.000000

In [ ]:
df_chases[df_chases['total_chases'] >= 20].sort_values(by='win_percentage', ascending=False)

team  total_chases  successfull_chases  \
6          Chennai Super Kings            75                  48   
3               Mumbai Indians            86                  48   
1        Kolkata Knight Riders            95                  51   
7          Sunrisers Hyderabad            45                  24   
5             Rajasthan Royals            79                  42   
0  Royal Challengers Bangalore            95                  45   
2             Delhi Daredevils            89                  40   
4              Kings XI Punjab            85                  37   
8              Deccan Chargers            32                  11   
9                Pune Warriors            25                   6   

   win_percentage  
6       64.000000  
3       55.813953  
1       53.684211  
7       53.333333  
5       53.164557  
0       47.368421  
2       44.943820  
4       43.529412  
8       34.375000  
9       24.000000

> **Insight:** Gujarat Lions has the highest success rate in chasing down targets with percentage of 75 (12 out of 16). But `if we consider atleast 20 chasing matches` then `Chennai Super Kings` has the best performance in chasing down targets with `48 wins out of 75` having 64% chance of win in chasing targets.

## **Q11**

In [5]:
df_ducks = df_ipl.groupby(['match_id', 'batsman'])[['batsman_runs']].sum().reset_index().sort_values(by='batsman_runs')
df_ducks = df_ducks[df_ducks['batsman_runs'] == 0].reset_index(drop=True)
df_ducks = df_ducks['batsman'].value_counts().sort_values(ascending=False)
df_ducks.columns = ['total_ducks']

In [7]:
df_ducks.head()

batsman
PA Patel           12
Harbhajan Singh    12
MK Pandey          12
RG Sharma          11
PP Chawla          11
Name: count, dtype: int64

> **Insight:** Players who have more ducks are `PA Patel, Harbhajan Singh, MK Pandey` with total ducks of 12.

## **Q12**

In [ ]:
def team_performance(df, team_name):
    df_1 = df[df['batting_team'] == team_name]
    df_2 = df[df['bowling_team'] == team_name]

    df_bat = df_1.groupby(['batting_team', 'bowling_team'])[['total_runs']].sum().sort_values(by='total_runs', ascending=False).reset_index()
    df_bat.columns = ['team_name', 'vs', 'total_runs']
    df_bowl = df_2.groupby(['bowling_team', 'batting_team'])[['player_dismissed']].count().reset_index()
    df_bowl.columns = ['team_name', 'vs', 'total_wickets']
    df_merged = pd.merge(df_bat, df_bowl, on=['team_name', 'vs'], how='inner')
    df = df_1.groupby(['match_id', 'bowling_team'])[['total_runs']].sum().reset_index()['bowling_team'].value_counts().reset_index()
    df_performance = pd.merge(df_merged, df, left_on='vs', right_on='bowling_team', how='inner')
    df_performance.rename(columns={'count': 'matches'}, inplace=True)
    df_performance.drop('bowling_team', axis=1, inplace=True)
    df_performance = df_performance.reindex(columns=['team_name', 'vs', 'matches', 'total_runs', 'total_wickets'])
    # for highest scores on each team
    df_scores = df_1.groupby(['match_id', 'bowling_team'])[['total_runs']].sum().reset_index()
    df_highest_scores = df_scores.groupby('bowling_team')[['total_runs']].max().reset_index()
    df_highest_scores.columns = ['vs', 'highest_score']
    df_performance = pd.merge(df_performance, df_highest_scores, on = 'vs', how='inner')

    return df_performance


In [ ]:
team_name = 'Royal Challengers Bangalore'
team_performance(df_ipl, team_name)

team_name                       vs  matches  total_runs  \
0   Royal Challengers Bangalore           Mumbai Indians       25        3993   
1   Royal Challengers Bangalore          Kings XI Punjab       24        3884   
2   Royal Challengers Bangalore    Kolkata Knight Riders       24        3622   
3   Royal Challengers Bangalore      Chennai Super Kings       23        3376   
4   Royal Challengers Bangalore         Delhi Daredevils       21        3274   
5   Royal Challengers Bangalore         Rajasthan Royals       20        2844   
6   Royal Challengers Bangalore      Sunrisers Hyderabad       14        2327   
7   Royal Challengers Bangalore          Deccan Chargers       11        1652   
8   Royal Challengers Bangalore            Pune Warriors        5         990   
9   Royal Challengers Bangalore            Gujarat Lions        5         934   
10  Royal Challengers Bangalore  Rising Pune Supergiants        2         380   
11  Royal Challengers Bangalore           Delhi Capitals        2         330   
12  Royal Challengers Bangalore     Kochi Tuskers Kerala        2         290   
13  Royal Challengers Bangalore   Rising Pune Supergiant        2         230   

    total_wickets  highest_score  
0             138            235  
1             135            226  
2             132            219  
3             134            213  
4             131            215  
5             105            204  
6              71            227  
7              67            185  
8              38            263  
9              34            248  
10             15            195  
11             11            176  
12             14            162  
13             11            134

## **Q13**

In [ ]:
df_mostruns_in_over_top5 = df_ipl.groupby(['match_id', 'inning', 'over'])[['total_runs']].sum().reset_index().sort_values(by='total_runs', ascending=False).head(5)
df_mostruns_in_over_top5.reset_index(drop=True, inplace=True)

In [ ]:
print('Below are the top 5 most runs overs:')
df_mostruns_in_over_top5

Below are the top 5 most runs overs:


match_id  inning  over  total_runs
0       283       2     3          37
1       208       1    13          33
2       516       2     6          33
3     11151       1    20          31
4       329       2    13          31

In [ ]:
mostruns_over = df_mostruns_in_over_top5.loc[0, 'over']
mostruns_over_id = df_mostruns_in_over_top5.loc[0, 'match_id']
mostruns_over_inning = df_mostruns_in_over_top5.loc[0, 'inning']
df_mostruns_in_over_details = df_ipl[(df_ipl['match_id'] ==  mostruns_over_id) & (df_ipl['inning'] ==  mostruns_over_inning) & (df_ipl['over'] == mostruns_over)].reset_index(drop=True)

In [ ]:
print('Below are the details of over with the highest number of runs scored:')
df_mostruns_in_over_details

Below are the details of over with the highest number of runs scored:


match_id  inning                 batting_team          bowling_team  over  \
0       283       2  Royal Challengers Bangalore  Kochi Tuskers Kerala     3   
1       283       2  Royal Challengers Bangalore  Kochi Tuskers Kerala     3   
2       283       2  Royal Challengers Bangalore  Kochi Tuskers Kerala     3   
3       283       2  Royal Challengers Bangalore  Kochi Tuskers Kerala     3   
4       283       2  Royal Challengers Bangalore  Kochi Tuskers Kerala     3   
5       283       2  Royal Challengers Bangalore  Kochi Tuskers Kerala     3   
6       283       2  Royal Challengers Bangalore  Kochi Tuskers Kerala     3   

   ball   batsman non_striker          bowler  is_super_over  ...  bye_runs  \
0     1  CH Gayle  TM Dilshan  P Parameswaran              0  ...         0   
1     2  CH Gayle  TM Dilshan  P Parameswaran              0  ...         0   
2     3  CH Gayle  TM Dilshan  P Parameswaran              0  ...         0   
3     4  CH Gayle  TM Dilshan  P Parameswaran              0  ...         0   
4     5  CH Gayle  TM Dilshan  P Parameswaran              0  ...         0   
5     6  CH Gayle  TM Dilshan  P Parameswaran              0  ...         0   
6     7  CH Gayle  TM Dilshan  P Parameswaran              0  ...         0   

   legbye_runs  noball_runs  penalty_runs  batsman_runs  extra_runs  \
0            0            0             0             6           0   
1            0            1             0             6           1   
2            0            0             0             4           0   
3            0            0             0             4           0   
4            0            0             0             6           0   
5            0            0             0             6           0   
6            0            0             0             4           0   

   total_runs  player_dismissed dismissal_kind fielder  
0           6               NaN            NaN     NaN  
1           7               NaN            NaN     NaN  
2           4               NaN            NaN     NaN  
3           4               NaN            NaN     NaN  
4           6               NaN            NaN     NaN  
5           6               NaN            NaN     NaN  
6           4               NaN            NaN     NaN  

[7 rows x 21 columns]

> **Insight:** In overall IPL matches, the over with highest runs scored is happenned in a match between `Royal Challengers Bangalore and Kochi Tuskers Kerala`. It is `CH Gayle from RCBangalore vs P Parameshwaran from KTKerala`. Total score of 37 runs with a spell of (6, 6+nb, 4, 4, 6, 6, 4).

## **Q14**

In [8]:
df_super_overs = df_ipl[df_ipl['is_super_over'] == 1]
df_runs_superover = df_super_overs.groupby(['match_id','inning', 'batting_team'])[['total_runs']].sum().reset_index()
df_runs_superover.head()

match_id  inning           batting_team  total_runs
0        34       3         Mumbai Indians          11
1        34       4          Gujarat Lions           6
2       126       3  Kolkata Knight Riders          15
3       126       4       Rajasthan Royals          18
4       190       3    Chennai Super Kings           9

In [ ]:
df_super_3 = df_runs_superover[df_runs_superover['inning'] == 3]
df_super_4 = df_runs_superover[df_runs_superover['inning'] == 4]
df_join = pd.merge(df_super_3, df_super_4, on=['match_id'], how='inner')
df_join['win'] = [f"{i[2]}" if i[3] > i[6] else f'{i[5]}' for i in df_join.values]

In [ ]:
print('Performance of teams in superovers:')
df_join.head()

Performance of teams in superovers:


match_id  inning_x               batting_team_x  total_runs_x  inning_y  \
0        34         3               Mumbai Indians            11         4   
1       126         3        Kolkata Knight Riders            15         4   
2       190         3          Chennai Super Kings             9         4   
3       388         3          Sunrisers Hyderabad            20         4   
4       401         3  Royal Challengers Bangalore            15         4   

                batting_team_y  total_runs_y                          win  
0                Gujarat Lions             6               Mumbai Indians  
1             Rajasthan Royals            18             Rajasthan Royals  
2              Kings XI Punjab            10              Kings XI Punjab  
3  Royal Challengers Bangalore            15          Sunrisers Hyderabad  
4             Delhi Daredevils            11  Royal Challengers Bangalore

In [ ]:
df_wins = df_join[['win']].value_counts().reset_index()
df_wins.columns = ['team', 'total wins']

In [ ]:
df_total = df_runs_superover['batting_team'].value_counts().reset_index()
df_superover_analysis = pd.merge(df_wins, df_total, left_on='team', right_on='batting_team', how='inner')
df_superover_analysis.drop('batting_team', axis=1, inplace=True)
df_superover_analysis.columns = ['team', 'total wins', 'total matches']
df_superover_analysis['win_percentage'] = (df_superover_analysis['total wins'] / df_superover_analysis['total matches']) * 100
df_superover_analysis

team  total wins  total matches  win_percentage
0              Kings XI Punjab           2              2      100.000000
1             Rajasthan Royals           2              3       66.666667
2               Mumbai Indians           1              1      100.000000
3  Royal Challengers Bangalore           1              2       50.000000
4          Sunrisers Hyderabad           1              1      100.000000

> Based on the above analysis, Kings XI Punjab given the best performance in superovers. But highest runs scored in superover is by Sunrisers Hyderabad with score of 20.

## **Q15**

In [9]:
def player_performance(df, player_name):
    batting_data = {'name': player_name, 'team': np.nan, 'matches': np.nan, 'total_runs': np.nan, 'balls_faced': np.nan, 'avg_score': np.nan,
                       'highest_score': np.nan, 'strike_rate': np.nan, '100s': np.nan, '50s': np.nan, '4s': np.nan, '6s': np.nan}
    bowling_data = {'name': player_name, 'team': np.nan, 'matches': np.nan,
                       'wickets': np.nan, 'overs': np.nan, 'runs_given': np.nan, 'avg_runs_given': np.nan,
                       'Best_bowling_inning': np.nan, 'economy': np.nan, 'strike_rate': np.nan, '4w': np.nan, '5w': np.nan}
    df_batting_stats = pd.DataFrame(batting_data, columns=batting_data.keys(), index=[0])
    df_bowling_stats = pd.DataFrame(bowling_data, columns=bowling_data.keys(), index=[0])
    df_batting = df[df['batsman'] == player_name]
    df_bowling = df[df['bowler'] == player_name]
    # batting stats
    df_batting_stats['team'] = df_batting['batting_team'].iloc[0]
    df_batting_stats['matches'] = df_batting['match_id'].nunique()
    df_batting_stats['total_runs'] = df_ipl[df_ipl['batsman'] == player_name]['batsman_runs'].sum()
    df_batting_stats['balls_faced'] = df_batting['ball'].count()
    df_batting_stats['avg_score'] = (df_batting_stats['total_runs'] / df_batting_stats['matches']).round(2)
    df_batting_stats['highest_score'] = df_batting.groupby('match_id')['batsman_runs'].sum().max()
    df_batting_stats['strike_rate'] = ((df_batting_stats['total_runs'] / df_batting_stats['balls_faced']) * 100).round(2)

    df_runs = df_batting.groupby('match_id')['batsman_runs'].sum()   # batsman runs in each match
    df_batting_stats['100s'] = df_runs[df_runs >= 100].count()
    df_batting_stats['50s'] = df_runs[(df_runs >= 50) & (df_runs < 100)].count()
    df_batting_stats['4s'] = df_batting[df_batting['batsman_runs'] == 4]['batsman_runs'].count()
    df_batting_stats['6s'] = df_batting[df_batting['batsman_runs'] == 6]['batsman_runs'].count()
    # return df_batting_stats

    # bowling stats
    df_bowling_stats['team'] = df_bowling['bowling_team'].iloc[0]
    df_bowling_stats['matches'] = df_bowling['match_id'].nunique()
    # counts of wickets each category
    df_wickets = df_bowling[(df_bowling['dismissal_kind'] != 'obstructing the field') & (df_bowling['dismissal_kind'] != 'retired hurt') &
                    (df_bowling['dismissal_kind'] != 'run out')]
    df_bowling_stats['wickets'] = df_wickets.value_counts('dismissal_kind').sum()
    df_bowling_stats['overs'] = df_bowling['ball'].count() / 6
    df_bowling_stats['runs_given'] = (df_bowling['total_runs'] - df_bowling['penalty_runs']).sum()
    df_bowling_stats['avg_runs_given'] = (df_bowling_stats['runs_given'] / df_bowling_stats['matches']).round(2)
    # calculating wickets taken in each match
    df_wickets_per_m = df_wickets.groupby('match_id')['dismissal_kind'].count()
    most_wickets_id = df_wickets_per_m[df_wickets_per_m == df_wickets_per_m.max()].index[0]
    df_bowling_stats['Best_bowling_inning'] = str(df_bowling[df_bowling['match_id'] == most_wickets_id]['total_runs'].sum()) + '/' + str(df_wickets_per_m.max())
    df_bowling_stats['economy'] = (df_bowling_stats['runs_given'] / df_bowling_stats['overs']).round(2)
    df_bowling_stats['strike_rate'] = df_bowling['ball'].count()/df_bowling_stats['wickets']
    df_bowling_stats['4w'] = (df_wickets.groupby('match_id')['dismissal_kind'].count() == 4).sum()
    df_bowling_stats['5w'] = (df_wickets.groupby('match_id')['dismissal_kind'].count() == 5).sum()

    return df_batting_stats, df_bowling_stats


In [10]:
player_name = 'V Kohli'
batting_stats, bowling_stats = player_performance(df_ipl, 'V Kohli')

In [11]:
print(f"Batting statistics of {player_name}:" )
batting_stats

Batting statistics of V Kohli:


name                         team  matches  total_runs  balls_faced  \
0  V Kohli  Royal Challengers Bangalore      169        5434         4211   

   avg_score  highest_score  strike_rate  100s  50s   4s   6s  
0      32.15            113       129.04     5   38  482  191

In [12]:
print(f"Bowling statistics of {player_name}:" )
bowling_stats

Bowling statistics of V Kohli:


name                         team  matches  wickets  overs  runs_given  \
0  V Kohli  Royal Challengers Bangalore       26        4   44.0         371   

   avg_runs_given Best_bowling_inning  economy  strike_rate  4w  5w  
0           14.27                25/2     8.43         66.0   0   0